duckduckgo/duckduckgo-privacy-extension
# Add Element Hiding Support

**Reviewer:** @jdorweiler 

**CC:** @jdorweiler @russellholt @kdzwinel @andrey-p 

## Description:
Hide iframes and their wrappers when they contain blocked requests. This adds functionality to deal with three common forms of ads:
1. Top level third-party iframe:
`<div class="ad-class">`	        <--- ad wrapper
	    `<iframe src="...">`		<--- blocked request
	    `</iframe>`
`</div>`

2. Nested third-party iframe
`<div class="ad-class">`	        <--- ad wrapper
	`<iframe src="...">`            <--- first party or otherwise unblocked request
		`<iframe src="...">`    <--- blocked request
		`</iframe>`
	`</iframe>`
`</div>`

3. Nested third-party scripts that would have loaded iframes
`<div class="ad-class">`	        <--- ad wrapper
	`<iframe src="...">`            <--- first party or otherwise unblocked request
		`<script src="...">`	<--- blocked request
                `</script>`
	`</iframe>`
`</div>`

I understand this code might be a bit hard to follow, since it relies on multiple content scripts running simultaneously in different iframes. See comments in element-hiding.js for more details. The rough description of what this does is:

1. Inject content scripts into all frames.
2. In each content script, wait for document to hit `readyState == 'interactive'`, which means the DOM in the frame is loaded (but scripts and other assets may not be yet).
3. a) If content script is running in main frame- gather list of top level iframes, store it, and send message to background page specifying that message is from main frame.
b) If content script is running in iframe - gather list of iframes and scripts contained within iframe, store it, and send message to background page specifying that message is from iframe.
4. In background page, listen for these messages from content scripts. If message is from main frame, respond with list of blocked requests of type frame. If message is from iframe, respond with list of blocked requests of type frame or type script.
5. In content scripts, listen for this message from background page and then:
a) If in main frame, iterate through the list of top level iframes, checking each src to see if it's included in the list of blocked frame requests. If it is, hide it and hide its parent node if the frame was the only child.
b) If in iframe, send message to content script in main frame containing list of blocked requests.
6. In main frame, listen for the message from 5b. When message is received, iterate through list of top level iframes and respond with the DOM id of each iframe element.
7. In each iframe, listen for message from 6 containing the frame's DOM id, then iterate through the list of found scripts and frames from 3b checking to see if any are in the list of blocked requests. If match is found, pass message to main frame containing the iframe's DOM id.
8. In main frame, listen for message containing frame id. When message is received, hide element with matching frame id and iterate upward recursively (if containers only contains iframe)


## Steps to test this PR:
1. Visit yahoo.com or other oath site where this is being tested
2. See that ads containing third party trackers are collapsed on page, but first party ads are not. finance.yahoo.com seems to consistently have first party ads that won't be hidden by this script. The recursive element hiding can also be tested on finance.yahoo.com -- the banner ad has a static container about five layers up from the iframe.
3. Visit random website, see that background page disables content scripts.

## Automated tests:

###### Reviewer Checklist:


###### PR Author Checklist:
